### PR TITLE
libbpfgo: make AttachTracepoint() signature 1:1 with libbpf

### PR DESCRIPTION
**Pull request by geyslan**
These changes, along with some refactoring, also update the `libbpfgo_test.go`.

Fixes:  #75

### CODE REVIEW

**Comment by rafaeldtinoco** [comment]
I have tested this with:

```C
SEC("tp/syscalls/sys_enter_sync")
int tracepoint__sys_enter_sync(struct trace_event_raw_sys_enter *args)
{
    return dealwithit(args);
}
```

```go
	// attach to BPF program to kprobe
	//_, err = bpfProgKsysSync.AttachTracepoint("syscalls:sys_enter_sync")
	_, err = bpfProgKsysSync.AttachTracepoint("syscalls", "sys_enter_sync")
	if err != nil {
		errExit(err)
	}
```

it works the same way (simplifying the split logic done within AttachTracepoint.

I think its okay to accept this (not that is changing anything meaningful but... it removes any doubts about category versus name.

**Comment by rafaeldtinoco** [comment]
**ORTHOGONAL** to this change: tracepoint attachment does not seem to be working in older kernels (not sure if because of what I predicted: tracepoint and perf events attachment share same baseline functions for attachment). 

> Not sure if super important also as we will likely drop support for those older kernels soon.

GO code:

```
$ sudo ./mine-static
libbpf: failed to parse target BTF: -2
libbpf: failed to perform CO-RE relocations: -2
libbpf: failed to load object 'mine.bpf.o'
2021/09/13 10:59:30 error: /home/rafaeldtinoco/work/sources/ebpf/mine-portablebpf/mine.go:79 failed to load BPF object
```
Equivalent C code:
```
$ sudo ./mine-c-static
Foreground mode...<Ctrl-C> or or SIG_TERM to end it.
libbpf: loading object 'mine_bpf' from buffer
libbpf: elf: section(3) tracepoint/syscalls/sys_enter_sync, size 320, link 0, flags 6, type=1
libbpf: sec 'tracepoint/syscalls/sys_enter_sync': found program 'tracepoint__sys_enter_sync' at insn offset 0 (0 bytes), code size 40 insns (320 bytes)
libbpf: elf: section(4) .reltracepoint/syscalls/sys_enter_sync, size 16, link 22, flags 0, type=9
libbpf: elf: section(5) license, size 4, link 0, flags 3, type=1
libbpf: license of mine_bpf is GPL
libbpf: elf: section(6) .maps, size 24, link 0, flags 3, type=1
libbpf: elf: section(13) .BTF, size 26786, link 0, flags 0, type=1
libbpf: elf: section(15) .BTF.ext, size 348, link 0, flags 0, type=1
libbpf: elf: section(22) .symtab, size 288, link 1, flags 0, type=2
libbpf: looking for externs among 12 symbols...
libbpf: collected 0 externs total
libbpf: map 'events': at sec_idx 6, offset 0.
libbpf: map 'events': found type = 4.
libbpf: map 'events': found key_size = 4.
libbpf: map 'events': found value_size = 4.
libbpf: sec '.reltracepoint/syscalls/sys_enter_sync': collecting relocation for section(3) 'tracepoint/syscalls/sys_enter_sync'
libbpf: sec '.reltracepoint/syscalls/sys_enter_sync': relo #0: insn #31 against 'events'
libbpf: prog 'tracepoint__sys_enter_sync': found map 0 (events, sec 6, off 0) for insn #31
libbpf: loading kernel BTF '/usr/lib/debug/boot/vmlinux-4.15.0-156-generic': 0
libbpf: Kernel doesn't support BTF, skipping uploading it.
libbpf: map 'events': setting size to 4
libbpf: map 'events': created successfully, fd=3
libbpf: sec 'tracepoint/syscalls/sys_enter_sync': found 2 CO-RE relocations
libbpf: CO-RE relocating [0] struct task_struct: found target candidate [142] struct task_struct in [vmlinux]
libbpf: CO-RE relocating [0] struct task_struct: found target candidate [25201] struct task_struct in [vmlinux]
libbpf: prog 'tracepoint__sys_enter_sync': relo #0: kind <byte_off> (0), spec is [19] struct task_struct.loginuid.val (0:123:0 @ offset 2992)
libbpf: prog 'tracepoint__sys_enter_sync': relo #0: matching candidate #0 [142] struct task_struct.loginuid.val (0:109:0 @ offset 2824)
libbpf: prog 'tracepoint__sys_enter_sync': relo #0: matching candidate #1 [25201] struct task_struct.loginuid.val (0:109:0 @ offset 2824)
libbpf: prog 'tracepoint__sys_enter_sync': relo #0: patched insn #15 (ALU/ALU64) imm 2992 -> 2824
libbpf: prog 'tracepoint__sys_enter_sync': relo #1: kind <byte_off> (0), spec is [19] struct task_struct.comm (0:102 @ offset 2792)
libbpf: prog 'tracepoint__sys_enter_sync': relo #1: matching candidate #0 [142] struct task_struct.comm (0:90 @ offset 2640)
libbpf: prog 'tracepoint__sys_enter_sync': relo #1: matching candidate #1 [25201] struct task_struct.comm (0:90 @ offset 2640)
libbpf: prog 'tracepoint__sys_enter_sync': relo #1: patched insn #22 (ALU/ALU64) imm 2792 -> 2640
Tracing... Hit Ctrl-C to end.
```

GO code works in newer kernels:

```
$ sudo ./mine-static
Listening for events, <Ctrl-C> or or SIG_TERM to end it.
sync (pid: 3932861) (loginuid: 1000)
```

File: libbpfgo.go [hunk 1]
```diff
@@ -752,25 +752,22 @@ func (p *BPFProg) SetTracepoint() error {
...
-	tpName := C.CString(tpEvent[1])
+func (p *BPFProg) AttachTracepoint(category, name string) (*BPFLink, error) {
+	tpCategory := C.CString(category)
+	defer C.free(unsafe.Pointer(tpCategory))
+	tpName := C.CString(name)
+	defer C.free(unsafe.Pointer(tpName))
+
 	link := C.bpf_program__attach_tracepoint(p.prog, tpCategory, tpName)
-	C.free(unsafe.Pointer(tpCategory))
-	C.free(unsafe.Pointer(tpName))
```
Comment (yanivagman) [inline libbpfgo.go:764]: nit: what was wrong with freeing the C strings here that we need to defer it?

File: libbpfgo.go [hunk 2]
```diff
@@ -752,25 +752,22 @@ func (p *BPFProg) SetTracepoint() error {
...
-		return nil, errptrError(unsafe.Pointer(link), "failed to attach tracepoint %s to program %s", tp, p.name)
+		return nil, errptrError(unsafe.Pointer(link), "failed to attach tracepoint %s to program %s", name, p.name)
 	}
 
 	bpfLink := &BPFLink{
 		link:      link,
 		prog:      p,
 		linkType:  Tracepoint,
-		eventName: tp,
+		eventName: fmt.Sprintf("%s:%s", category, name),
```
Comment (yanivagman) [inline libbpfgo.go:770]: I think that it is ok if we just use `name` here so it will match the above error message where we also return `name`

**Comment by yanivagman** [comment]
> ORTHOGONAL to this change: tracepoint attachment does not seem to be working in older kernels (not sure if because of what I predicted: tracepoint and perf events attachment share same baseline functions for attachment).
> 
> Not sure if super important also as we will likely drop support for those older kernels soon.
> 
> GO code:
> 
> ```
> $ sudo ./mine-static
> libbpf: failed to parse target BTF: -2
> libbpf: failed to perform CO-RE relocations: -2
> libbpf: failed to load object 'mine.bpf.o'
> 2021/09/13 10:59:30 error: /home/rafaeldtinoco/work/sources/ebpf/mine-portablebpf/mine.go:79 failed to load BPF object
> ```

@rafaeldtinoco I'm not sure that this problem has anything to do with tracepoints. It seems to me more of a CO-RE problem (libbpf: failed to parse target BTF: -2). I believe that your `mine-static` program is CO-RE, while you are running on a kernel which doesn't support BTF (as can be seen in the C equivalent code: `libbpf: Kernel doesn't support BTF, skipping uploading it.`). Can you please try to compile it "non-CO-RE" and try again?

File: libbpfgo.go (hunk 1 above)
Comment (geyslan) [reply libbpfgo.go:764]: It's just a nitpicking, perhaps trying to use the Golang idioms. But this change is not mandatory. If rejected I can get rid of it.

File: libbpfgo.go (hunk 2 above)
Comment (geyslan) [reply libbpfgo.go:770]: I see. Consider it done.

File: libbpfgo.go (hunk 1 above)
Comment (yanivagman) [reply libbpfgo.go:764]: Well in this case, we can just free the allocated objects right after we use them, and it seems to be clearer (at least for me).
I would use defer when there is an operation (like closing a file) that can't happen right away, and I want to ensure that will happen in any return path of the function.

**Review by yanivagman — APPROVED** [review]

**Pull request merged**